### PR TITLE
[object] Object runtime and state refactor

### PR DESCRIPTION
**Pull request by jolestar**
## Summary

1. Remove `ObjectEntity<T>` from `object.move` module, and maintain ObjectMeta in object runtime.
2. Implement `object::transfer`, `object::to_shared`, `object::to_frozen` in native function.
3. Add the `store` ability required to the table key.
4. Add some native functions to read ObjectMeta.
5. Make dynamic field to Object<DynamicField<Name,Value>>, all State are Object now.
6. Unify the field key to address.
7. Rename `State` to `ObjectState`. All states are now Objects. 
8. Make the `state_root` field optional to optimize the storage. 

Please help review the object runtime and `object.move` @baichuan3 @steelgeek091 @pause125 @wow-sven 

The following issues will be fixed in the new PR.

- [ ] Migrate `Timestamp` to timestamp.move
- [ ] Define Timestamp and ModuleStore object in genesis config.
- [ ] Fix the Object and resource display, How to access the metadata in the template.
- [ ] Fix import and export tools.
- [ ] Simply the root metadata in StateChangeSet.
- [ ] Unify `object_type` of ObjectStateView and `value_type` of ObjectState field
- [ ] Unify fields of IndexerObjectStateView, ObjectStateView, ObjectMetaView, StateKVView(key_hex rename to field_key?)
- [ ] Update the state_root of ObjectMeta in StateChangeSet to the latest before returning to users.
- [ ] Directly use block hash as the PendingBlockID
- [ ] Indexer update for Object metadata change

part of #2024

### CODE REVIEW

File: moveos/moveos-object-runtime/src/runtime_object.rs [hunk 1]
```diff
@@ -81,533 +57,435 @@ impl ObjectPointer {
...
             ObjectPointer::none()
         } else {
             ObjectPointer::cached(id.clone())
         };
         Ok(Self {
-            id: id.clone(),
-            value_layout,
-            value_type: state.value_type,
+            rt_meta: RuntimeObjectMeta::cached(metadata, value_layout),
             value: GlobalValue::cached(value)?,
```
Comment (baichuan3) [inline moveos/moveos-object-runtime/src/runtime_object.rs:101]: Object Meta和Object Value 全局cache是在哪缓存的？

Comment (jolestar) [reply moveos/moveos-object-runtime/src/runtime_object.rs:101]: 全局 cache 还没加，当前只重构了 runtime 和 native function 部分，statedb 还没改。

File: moveos/moveos-types/src/moveos_std/object/dynamic_field.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    tracing::trace!(
+        "Deriving dynamic field key for key={:?}, key_type_tag={:?}",
+        key,
+        k_tag_str,
+    );
+
+    // hash(key || key_type_tag)
+    let mut buffer = bcs::to_bytes(key)?;
+    buffer.extend_from_slice(k_tag_str.as_bytes());
+    let field_key = h256::sha3_256_of(&buffer);
```
Comment (baichuan3) [inline moveos/moveos-types/src/moveos_std/object/dynamic_field.rs:101]: 重构后所有的 DynamicField 都是 Object，之前担心的 ObjectID 可能重复的概率会增加，有办法在 framework 里感知到 objectID 和 address 冲突吗？

Comment (jolestar) [reply moveos/moveos-types/src/moveos_std/object/dynamic_field.rs:101]: DynamicField 的 key 是通过 key + 类型生成的，在当前 parent 下保持唯一就行。因为 Root Object 不支持直接添加 DynamicField，所以不会和其他 Object 冲突。Child Object 的 ObjectID 是一个 path，也不会有问题。
现在可能会有问题的是 custom object id，那个写了个 TODO。需要再考虑一下。

**Review by baichuan3 — APPROVED** [review]

File: moveos/moveos-types/src/moveos_std/object/dynamic_field.rs (hunk 2 above)
Comment (jolestar) [reply moveos/moveos-types/src/moveos_std/object/dynamic_field.rs]: Field key 和 custom Object ID 通过两个方法分别处理了

https://github.com/rooch-network/rooch/blob/0bed3c44ac61293387943528299f767ddf2b5b36/frameworks/moveos-stdlib/sources/object.move#L593-L613

File: crates/rooch/src/commands/statedb/commands/genesis_ord.rs [hunk 3]
```diff
@@ -472,13 +472,17 @@ fn produce_ord_updates(tx: SyncSender<BatchUpdatesOrd>, input: PathBuf, batch_si
...
-
-    let state = inscription_id.into_state();
-
-    (key_state, state)
+fn gen_inscription_ids_update(
+    index: u64,
+    inscription_id: InscriptionID,
+) -> (FieldKey, ObjectState) {
+    //let key = bcs::to_bytes(&index).expect("bcs to_bytes u64 must success.");
+    //TODO we need to get the TableVec object id from args
```
Comment (popcnt1) [inline crates/rooch/src/commands/statedb/commands/genesis_ord.rs:480]: Shouldn't it be Table, not TableVec, as the parent id?

File: frameworks/moveos-stdlib/sources/object.move [hunk 4]
```diff
@@ -437,270 +341,174 @@ module moveos_std::object {
...
+        let Object{id:_} = obj;
     }
 
-    #[private_generics(T, V)]
+    #[private_generics(T, Value)]
     /// Add a object field to the object. return the child object
     /// The parent object must be a shared object
-    public fun add_object_field<T: key, V: key>(obj: &mut Object<T>, v: V): Object<V> {
+    public fun add_object_field<T: key, Value: key>(obj: &mut Object<T>, v: Value): Object<Value> {
         // Only shared object can add child object
```
[the remark this answers is not in the document]
Comment (jolestar) [reply frameworks/moveos-stdlib/sources/object.move:370]: 我也在想。干脆改成 `new_with_parent(parent:&mut Object<T>, child_value: V): Object<T>` 这样？

File: crates/rooch/src/commands/statedb/commands/genesis_ord.rs (hunk 3 above)
Comment (jolestar) [reply crates/rooch/src/commands/statedb/commands/genesis_ord.rs:480]: It is `inscriptions: TableVec<InscriptionID>` field of InscriptionStore

File: frameworks/moveos-stdlib/sources/object.move (hunk 4 above)
[the remark this answers is not in the document]
Comment (popcnt1) [reply frameworks/moveos-stdlib/sources/object.move:370]: > 我也在想。干脆改成 `new_with_parent(parent:&mut Object<T>, child_value: V): Object<T>` 这样？

fork？现在结构和进程树一样了

File: crates/rooch/src/commands/statedb/commands/genesis_ord.rs (hunk 3 above)
Comment (popcnt1) [reply crates/rooch/src/commands/statedb/commands/genesis_ord.rs:480]: > It is `inscriptions: TableVec<InscriptionID>` field of InscriptionStore

Inside TableVec is a Table, my original implementation:

1. put index:inscription in Table
2. using Table make TableVec

File: frameworks/moveos-stdlib/sources/object.move (hunk 4 above)
[the remark this answers is not in the document]
Comment (jolestar) [reply frameworks/moveos-stdlib/sources/object.move:370]: 好像也是个思路。

File: crates/rooch/src/commands/statedb/commands/genesis_ord.rs (hunk 3 above)
Comment (jolestar) [reply crates/rooch/src/commands/statedb/commands/genesis_ord.rs:480]: yes. need to pass the id here

**Comment by jolestar** [comment]
@wow-sven RPC 返回值有变化，sdk 我尝试修复了一下。

File: frameworks/moveos-stdlib/sources/object.move (hunk 4 above)
[the remark this answers is not in the document]
Comment (jolestar) [reply frameworks/moveos-stdlib/sources/object.move:370]: use `new_with_parent`

**Pull request merged**